iroh-gossip/src/proto/plumtree.rs

```diff
@@ -0,0 +1,462 @@
+    /// included message ID.
+    Graft(Graft),
+    /// When receiving IHave, do nothing initially, and request the messages for the included
+    /// message IDs after some time if they aren't pushed eagerly to us.
+    IHave(Vec<IHave>),
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Gossip {
+    id: MessageId,
+    round: Round,
+    content: Bytes,
+}
+impl Gossip {
+    pub fn next_round(self) -> Gossip {
+        Gossip {
+            id: self.id,
+            content: self.content,
+            round: self.round.next(),
+        }
+    }
+}
+impl fmt::Debug for Gossip {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Gossip")
+            .field("id", &self.id)
+            .field("round", &self.round)
+            .field("content", &format!("<{}>", self.content.len()))
+            .finish()
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct IHave {
+    id: MessageId,
+    round: Round,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct Graft {
+    id: Option<MessageId>,
+    round: Round,
+}
+
+#[derive(Clone, Debug)]
+pub struct Config {
+    pub graft_timeout_1: Duration,
+    pub graft_timeout_2: Duration,
+    pub dispatch_timeout: Duration,
+    pub optimization_threshold: Round,
+}
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            graft_timeout_1: Duration::from_millis(80),
+            graft_timeout_2: Duration::from_secs(40),
+            dispatch_timeout: Duration::from_millis(50),
+            optimization_threshold: Round(7),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct Stats {
+    pub payload_messages_received: u64,
+    pub control_messages_received: u64,
+    pub max_last_delivery_hop: u16,
+}
+
+#[derive(Debug)]
+pub struct State<PA> {
+    me: PA,
+    config: Config,
+
+    pub(crate) eager_push_peers: HashSet<PA>,
+    pub(crate) lazy_push_peers: HashSet<PA>,
+
+    lazy_push_queue: HashMap<PA, Vec<IHave>>,
+
+    missing_messages: HashMap<MessageId, VecDeque<(PA, Round)>>,
+    received_messages: IndexSet<MessageId>,
+    cache: HashMap<MessageId, Gossip>,
+
+    graft_timer_scheduled: HashSet<MessageId>,
+    dispatch_timer_scheduled: bool,
+
+    pub(crate) stats: Stats,
+}
+
+impl<PA: PeerAddress> State<PA> {
+    pub fn new(me: PA, config: Config) -> Self {
+        Self {
+            me,
+            eager_push_peers: Default::default(),
+            lazy_push_peers: Default::default(),
+            lazy_push_queue: Default::default(),
+            config,
+            missing_messages: Default::default(),
+            received_messages: Default::default(),
+            graft_timer_scheduled: Default::default(),
+            cache: Default::default(),
+            dispatch_timer_scheduled: false,
+            stats: Default::default(),
+        }
+    }
+
+    // TODO: optimization step from the paper
```

Comment:
how important is this?

Reply:
It's actually already implemented, forgot to remove the TODO.